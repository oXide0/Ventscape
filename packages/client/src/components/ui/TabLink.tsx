import { Link as ChakraLink, Stack, Text, useTheme } from '@chakra-ui/react';
import { IconContext } from 'react-icons';
import { Link as ReactRouterLink, useMatch } from 'react-router-dom';

interface TabLinkProps {
    to: string;
    icon: React.ReactNode;
    label: string;
    isBold?: boolean;
}

const TabLink = ({ to, icon, label, isBold }: TabLinkProps) => {
    const isEditProfileActive = useMatch('/profile/edit');
    const match = useMatch(to);
    const theme = useTheme();
    const isActive = (to === '/profile' && isEditProfileActive) || match;

    return (
        <IconContext.Provider
            value={{
                size: '1.7em',
                color: isActive ? theme.colors.text.primary : theme.colors.text.secondary,
            }}
        >
            <ChakraLink as={ReactRouterLink} to={to} style={{ textDecoration: 'none' }}>
                <Stack
                    direction='row'
                    alignItems='center'
                    justifyContent={{ base: 'center', xl: 'flex-start' }}
                    _hover={{ bg: 'bg.default' }}
                    _active={{ transform: 'scale(0.95)', transition: 'all 0.2s ease-in-out' }}
                    bg={isActive ? 'bg.default' : 'transparent'}
                    px={3}
                    py={2}
                    rounded='md'
                    cursor='pointer'
                >
                    {icon}
                    <Text
                        color={isActive ? 'text.primary' : 'text.secondary'}
                        display={{ base: 'none', xl: 'block' }}
                        fontWeight={isBold ? 'bold' : 'normal'}
                    >
                        {label}
                    </Text>
                </Stack>
            </ChakraLink>
        </IconContext.Provider>
    );
};

export default TabLink;
