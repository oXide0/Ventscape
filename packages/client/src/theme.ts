import {
    extendTheme,
    ThemeConfig,
    defineStyleConfig,
    createMultiStyleConfigHelpers,
} from '@chakra-ui/react';
import { cardAnatomy } from '@chakra-ui/anatomy';

const { definePartsStyle, defineMultiStyleConfig } = createMultiStyleConfigHelpers(
    cardAnatomy.keys
);

const config: ThemeConfig = {
    initialColorMode: 'dark',
    useSystemColorMode: false,
};

const colors = {
    bg: {
        default: '#252B3C',
        navbar: '#2F3647',
    },
    text: {
        primary: '#FFFFFF',
        secondary: '#A0AEC0',
    },
    brand: {
        100: '#6B72FF',
        200: '#5056ED',
    },
};

const styles = {
    global: {
        body: {
            bg: 'bg.default',
            color: 'text.primary',
            overflowX: 'hidden',
        },
        '::-webkit-scrollbar': {
            width: '10px',
        },
        '::-webkit-scrollbar-track': {
            background: 'none',
        },
        '::-webkit-scrollbar-thumb': {
            background: 'text.secondary',
            borderRadius: '4px',
        },
        '::-webkit-scrollbar-thumb:hover': {
            background: 'brand.100',
        },
    },
};

const Spinner = defineStyleConfig({
    baseStyle: {
        color: 'purple.400',
        position: 'absolute',
        top: '45%',
        left: '48%',
    },
    defaultProps: {
        size: 'xl',
    },
});
const Divider = defineStyleConfig({
    baseStyle: {
        bg: 'text.secondary',
    },
});
const Text = defineStyleConfig({
    baseStyle: {
        color: 'text.primary',
    },
});

const Input = defineStyleConfig({
    defaultProps: {
        // @ts-expect-error This is intentional to set a default focusBorderColor.
        focusBorderColor: 'brand.100',
    },
});
const Select = defineStyleConfig({
    defaultProps: {
        // @ts-expect-error This is intentional to set a default focusBorderColor.
        focusBorderColor: 'brand.100',
    },
});

const baseStyle = definePartsStyle({
    container: {
        backgroundColor: 'bg.navbar',
        borderRadius: '5px',
    },
    header: {
        backgroundColor: 'bg.navbar',
        borderRadius: '5px',
    },
    body: {
        backgroundColor: 'bg.navbar',
        borderRadius: '5px',
    },
    footer: {
        backgroundColor: 'bg.navbar',
        borderRadius: '5px',
    },
});

export const cardTheme = defineMultiStyleConfig({ baseStyle });

const theme = extendTheme({
    config,
    styles,
    colors,
    components: { Spinner, Divider, Text, Input, Select, Card: cardTheme },
});

export default theme;
