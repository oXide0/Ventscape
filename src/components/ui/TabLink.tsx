import { Stack, Link as ChakraLink, Text, useTheme } from '@chakra-ui/react';
import { IconContext } from 'react-icons';
import { Link as ReactRouterLink, useMatch } from 'react-router-dom';

interface TabLinkProps {
    to: string;
    icon: React.ReactNode;
    label: string;
    isCollapsed?: boolean;
}

const TabLink = ({ to, icon, label, isCollapsed }: TabLinkProps) => {
    const match = useMatch(to);
    const theme = useTheme();

    return (
        <IconContext.Provider
            value={{
                size: '1.7em',
                color: match ? theme.colors.text.primary : theme.colors.text.secondary,
            }}
        >
            <ChakraLink as={ReactRouterLink} to={to} style={{ textDecoration: 'none' }}>
                <Stack
                    direction='row'
                    alignItems='center'
                    justifyContent={{ base: 'center', xl: isCollapsed ? 'center' : 'flex-start' }}
                    _hover={{ bg: 'bg.default' }}
                    _active={{ transform: 'scale(0.95)', transition: 'all 0.2s ease-in-out' }}
                    bg={match ? 'bg.default' : 'transparent'}
                    px={3}
                    py={2}
                    rounded='md'
                    cursor='pointer'
                >
                    {icon}
                    {isCollapsed ? null : (
                        <Text
                            color={match ? 'text.primary' : 'text.secondary'}
                            display={{ base: 'none', xl: 'block' }}
                        >
                            {label}
                        </Text>
                    )}
                </Stack>
            </ChakraLink>
        </IconContext.Provider>
    );
};

export default TabLink;
