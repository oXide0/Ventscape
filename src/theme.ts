import { extendTheme, ThemeConfig } from '@chakra-ui/react';

const config: ThemeConfig = {
    initialColorMode: 'dark',
    useSystemColorMode: false,
};

const colors = {
    bg: {
        default: '#252B3C',
        navbar: '#2F3647',
    },
    text: {
        primary: '#FFFFFF',
        secondary: '#A0AEC0',
    },
    brand: {
        100: '#6B72FF',
        200: '#5056ED',
    },
};

const styles = {
    global: {
        body: {
            bg: 'bg.default',
            color: 'text.primary',
            overflowX: 'hidden',
        },
        '::-webkit-scrollbar': {
            width: '10px',
        },
        '::-webkit-scrollbar-track': {
            background: 'none',
        },
        '::-webkit-scrollbar-thumb': {
            background: 'text.secondary',
            borderRadius: '4px',
        },
        '::-webkit-scrollbar-thumb:hover': {
            background: 'brand.100',
        },
    },
};

const theme = extendTheme({ config, styles, colors });

export default theme;
