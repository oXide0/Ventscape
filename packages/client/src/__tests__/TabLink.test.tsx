import { render } from '@testing-library/react';
import TabLink from 'components/ui/TabLink';
import { FaHome } from 'react-icons/fa';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from '@chakra-ui/react';
import theme from '../theme';

describe('TabLink component', () => {
    it('renders without crashing', () => {
        render(
            <ThemeProvider theme={theme}>
                <MemoryRouter>
                    <TabLink to='/' icon={<FaHome />} label='Home' />
                </MemoryRouter>
            </ThemeProvider>
        );
    });

    it('renders label correctly', () => {
        const label = 'Home';
        const { getByText } = render(
            <ThemeProvider theme={theme}>
                <MemoryRouter>
                    <TabLink to='/' icon={<FaHome />} label={label} />
                </MemoryRouter>
            </ThemeProvider>
        );
        expect(getByText(label)).toBeInTheDocument();
    });

    it('creating snapshot for label', () => {
        const { getByText } = render(
            <ThemeProvider theme={theme}>
                <MemoryRouter>
                    <TabLink to='/' icon={<FaHome />} label='Home' />
                </MemoryRouter>
            </ThemeProvider>
        );
        expect(getByText('Home')).toMatchSnapshot();
    });
});
