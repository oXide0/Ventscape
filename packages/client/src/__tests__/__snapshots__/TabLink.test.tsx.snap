// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabLink component creating snapshot for label 1`] = `
<p
  class="chakra-text css-hu7h93"
>
  Home
</p>
`;
